proposals/src/proposal/cosmwasm/anchor_update.rs

```diff
@@ -0,0 +1,238 @@
+            src_chain_id: self.src_chain.chain_id(),
+            root: self.merkle_root,
+            latest_leaf_index: self.latest_leaf_index,
+            target: self.target,
+        })
+        .unwrap();
+
+        bytes.extend_from_slice(&message.to_vec());
+
+        bytes
+    }
+
+    /// Get the proposal as a bytes without copying.
+    #[must_use]
+    pub fn into_bytes(self) -> Vec<u8> {
+        self.to_bytes()
+    }
+}
+
+impl From<Vec<u8>> for AnchorUpdateProposal {
+    fn from(bytes: Vec<u8>) -> Self {
+        let f = 0usize;
+        let t = ProposalHeader::LENGTH;
+        let mut header_bytes = [0u8; ProposalHeader::LENGTH];
+        header_bytes.copy_from_slice(&bytes[f..t]);
+        let header = ProposalHeader::from(header_bytes);
+
+        let f = t;
+        let msg_bytes = bytes[f..].to_vec();
+        let decoded_edge_data: UpdateEdge = from_slice(&msg_bytes).unwrap();
+
+        Self::new(
+            header,
+            TypedChainId::from(decoded_edge_data.src_chain_id),
+            decoded_edge_data.latest_leaf_index,
+            decoded_edge_data.root,
+            decoded_edge_data.target,
+        )
+    }
+}
+
+impl From<AnchorUpdateProposal> for Vec<u8> {
+    fn from(proposal: AnchorUpdateProposal) -> Self {
+        proposal.to_bytes()
+    }
+}
+
+// if we have EVM available, we can convert the EVM proposal to a cosmwasm proposal
+#[cfg(feature = "evm")]
+impl From<crate::evm::AnchorUpdateProposal> for AnchorUpdateProposal {
+    fn from(proposal: crate::evm::AnchorUpdateProposal) -> Self {
+        AnchorUpdateProposal::new(
+            proposal.header(),
+            proposal.src_chain(),
+            proposal.latest_leaf_index(),
+            *proposal.merkle_root(),
+            *proposal.target(),
+        )
+    }
+}
+
+// if we have Substrate available, we can convert the EVM proposal to a cosmwasm proposal
+#[cfg(feature = "substrate")]
+impl From<crate::substrate::AnchorUpdateProposal> for AnchorUpdateProposal {
+    fn from(proposal: crate::substrate::AnchorUpdateProposal) -> Self {
+        AnchorUpdateProposal::new(
+            proposal.header(),
+            proposal.src_chain(),
+            proposal.latest_leaf_index(),
+            *proposal.merkle_root(),
+            *proposal.target(),
+        )
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+struct UpdateEdge {
+    src_chain_id: u64,
+    root: [u8; 32],
+    latest_leaf_index: u32,
+    target: [u8; 32],
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        cosmwasm::cosmos_addr_2_target_addr, FunctionSignature, Nonce,
+        ResourceId, TargetSystem,
+    };
+
+    use super::*;
+
+    const TARGET_CONTRACT_ADDR: &str =
+        "juno1hset4pny4h8xm4s4lek57msq7j4zwfqwjf7zxqjt4npxyv0lrgnsp8qy9j";
+
+    #[test]
+    fn encode() {
+        let target_addr = cosmos_addr_2_target_addr(TARGET_CONTRACT_ADDR);
```

Comment:
Can we rename this to `cosmos_address_to_target_system`?

Reply:
Um, atm, it is for converting the `cosmos address` to **20 bytes** hashed address, not the `target_system`(26 bytes).
That's why I named it as `cosmos_addr_2_target_addr`.

Reply:
Can you rename it to `cosmos_address_to_target_address` then please? I don't particularly like mixing numbers to the functions when it's meant as `to`.

Reply:
Done https://github.com/webb-tools/webb-rs/pull/64/commits/1c456ad3f6d57b4b4a9a2e92c406ef7932a1f7c8